### ISSUE TITLE
Templates for grype output. HTML template

### ISSUE THREAD

**Issue by myrkytyn** [anchore/grype#724]
Hi! Does someone have templates for grype? I need to scan my images to human-readable format HTML.

I think will be great if we can have a directory with templates for grype project.

**Comment by freedom-isnotanarchy** [comment]
(1) ((directory with templates)): Yes, great idea !
(2) Here is 1 grype template that I use:
`"0-FILE:"#"PACKAGE:"#"VULNERABILITY:"#"SEVERITY:"#"Is Fixed:"#"Version Found:"#"Version Fixed:"#"Type:"#"Namespace:"#"Package URL:"#"Vulnerability Source:"#"Description:"#"URLS:"
{{- range .Matches}}
"__FILENAME__"#"{{.Artifact.Name}}"#"{{.Vulnerability.ID}}"#"{{.Vulnerability.Severity}}"#"{{.Vulnerability.Fix.State}}"#"'{{.Artifact.Version}}"#"{{.Vulnerability.Fix.Versions}}"#"{{.Artifact.Type}}"#"{{.Vulnerability.Namespace}}"#"{{.Artifact.PURL}}"#"{{.Vulnerability.DataSource}}"#"{{.Vulnerability.Description}}"#"{{.Vulnerability.URLs}}"
{{- end}}`

**Comment by samcro1967** [comment]
Is the correct usage to redirect the output to an html file for viewing in a browser in a human readable format?  It generates the html files with the output, but it is not formatted in any way.  Guessing I a not doing this correctly or am missing something.

```
grype -c grype.yaml -o template -t custom.template --scope all-layers --only-fixed sbom:./sboms/$CONTAINER.json > ./results/${TIMESTAMP}_$CONTAINER.html
```

![image](https://user-images.githubusercontent.com/2319445/163690019-1aba4cc8-3825-4ea8-85ce-6f0608031ae6.png)

**Comment by adriens** [comment]
Sorry for asking here, but it's also template related.

Is there any `markdown` ready to use template ?... I'd like to psot-process it with `pandoc` to build a ready to use automated documentation toolchain... human redable.

BTW : that would fix the `html` generation.

**Comment by ida-greuelr** [comment]
Inspired from the html output of https://sourceforge.net/projects/command-output-to-html-table/ I created a simple html table template. Works fine for me, is human readable and interactive:
```html
<!DOCTYPE html>
<html>

<head>
  <meta name="viewport" content="width=device-width, initial-scale=1" charset="ISO-8859-1">
  <title>Grype</title>
  <script>
    function myFunction() {
      // Declare variables 
      var input, filter, table, tr, td, i;
      input = document.getElementById("myInput");
      filter = input.value.toUpperCase();
      table = document.getElementById("myTable");
      tr = table.getElementsByTagName("tr");

      // Loop through all table rows, and hide those who do not match the search query
      for (i = 1; i < tr.length; i++) {
        td = tr[i].getElementsByTagName("td");
        for (j = 0; j < td.length; j++) {
          var tdata = td[j];
          if (tdata) {
            if (tdata.innerHTML.toUpperCase().indexOf(filter) > -1) {
              tr[i].style.display = "";
              break;
            } else {
              tr[i].style.display = "none";
            }
          }
        }
      }
    }
  </script>

  <style type="text/css">
    table,
    th,
    td {
      border: 1px solid black; // changing-colors
      // word-wrap: break-word;
    }

    tr:first-child {
      font-weight: bold;
    }

    tr:nth-child(even) {
      background: #eef
    }

    // changing-colors
    tr:nth-child(odd) {
      background: #fee
    }

    // changing-colors

    * {
      box-sizing: border-box;
    }

    #myInput {
      background-image: url("search.png");
      background-position: left center;
      background-repeat: no-repeat;
      width: 94%;
      font-size: 16px;
      padding: 8px 20px 8px 40px;
      border: 2px solid Tomato; // changing-colors
      margin-bottom: 18px;
    }

    #myTable {
      border-collapse: collapse;
      border: 1px solid #ddd; // changing-colors
      width: 100%;
      margin-top: 18px;
      // Remove the // in front of the below two lines, to get fixed-width
      // table-layout: fixed;
      // word-wrap: break-word;
      // font-size: 18px;
    }

    #myTable th,
    #myTable td {
      text-align: left;
      padding: 12px;
    }

    #myTable tr {
      border-bottom: 1px solid #ddd; // changing-colors
    }

    #myTable tr:first-child:hover,
    #myTable tr:hover {
      background-color: rgb(4, 112, 155); // changing-colors
    }

    #myTable tr:first-child {
      background-color: rgb(4, 112, 155); // changing-colors
      font-weight: bold;
    }
  </style>
</head>

<body>
  <h2 style="text-align:center;background-color:DodgerBlue;color:White;">Identified Vulnerabilites</h2>
  <input type="text" id="myInput" onkeyup="myFunction()" placeholder="Search in all Fields...."
    title="Type in a Search String" autofocus="autofocus">
  <table id="myTable">
    <tr>
      <th>NAME</th>
      <th>INSTALLED</th>
      <th>TYPE</th>
      <th>VULNERABILITY</th>
      <th>SEVERITY</th>
      <th>DESCRIPTION</th>
      <th>STATE</th>
      <th>FIXED IN</th>
    </tr>
{{- range .Matches}}
    <tr>
      <td>"{{.Artifact.Name}}"</td>
      <td>"{{.Artifact.Version}}"</td>
      <td>"{{.Artifact.Type}}"</td>
      <td>"{{.Vulnerability.ID}}"</td>
      <td>"{{.Vulnerability.Severity}}"</td>
      <td>"{{.Vulnerability.Description}}"</td>
      <td>"{{.Vulnerability.Fix.State}}"</td>
      <td>"{{.Vulnerability.Fix.Versions}}"</td>
    </tr>
{{- end}}
  </table>
</body>

</html>
```
Copy/paste into a file like `html.tmpl` and run for example:
```bash
grype docker:ubuntu:focal -o template -t html.tmpl > grype.html
```
Praise the sourceforge project for its html layout :1st_place_medal:  !

**Comment by spiffcs** [comment]
@ida-greuelr this is awesome!

**Comment by spiffcs** [comment]
Closing this issue with the template provided if anyone else objects or has related pr to file under issue just tag me and we can sort it

**Comment by spiffcs** [comment]
Closing this issue with the template provided if anyone else objects or has related pr to file under issue just tag me and we can sort it

**Comment by adriens** [comment]
Hi, can I access the name to the input scanned object (docker image, file, directory,...) through a variable from within the template :grey_question:  That would be very usefiul.

Should I create a dedicated issue ?

**Comment by adriens** [comment]
Forget my previous question cf [documentation](https://github.com/anchore/grype#using-templates) : 


> Grype's template processing uses the same data models as the json output format — so if you're wondering what data is available as you author a template, you can use the output from grype <image> -o json as a reference.

**Comment by adriens** [comment]
@ida-greuelr , to put the template in security and start some mode devs on templates, I created this https://dev.to/optnc/grype-0420-is-out-and-hello-grype-contribs-3g4i

We'll work on imprving templates and some other stuffs.

**Comment by rtarquini** [comment]
I'm having trouble understanding how the json names map back to the template. I was trying to extract the artifact path using .Artifact.Locations.Path, but that throws the error below

 _ERROR unable to show grype-vulnerability-scanning-finished event: unable to show vulnerability report: unable to execute supplied template: template: html.tmpl:131:22: executing "html.tmpl" at <.Artifact.Locations.Path>: can't evaluate field Path in type []source.Coordinates_

**Comment by kzantow** [comment]
@rtarquini `.Artifact.Locations` is a slice (list/array), so you'll need to use a loop construct, I think it would be something like:

```
{{- range .Artifact.Locations}}
  {{.Path}}
{{- end}}
```

**Comment by mike-19** [comment]
@kzantow  `{{.Path}}` will not work and will result in the same error; this is primarily due to improper definitions in the data model  referenced here in https://github.com/anchore/grype/issues/963


```
{{- range .Artifact.Locations}}
  {{(index .Artifact.Locations 0).Path}}
{{- end}}

```

**Comment by kzantow** [comment]
@mike-19 ah, I forgot about that `RealPath` detail. But note you would only want to use `{{(index .Artifact.Locations 0).Path}}` if you are just getting the first element, the `{{- range...` stuff is for iterating over each element. So if you just want the first element, you can use: `{{(index .Artifact.Locations 0).RealPath}}` if you want _all_ locations, you can use `{{- range $i, $e := .Artifact.Locations}}{{- if $i}}, {{- end}}{{$e.RealPath}}{{- end}}`

**Comment by OnceUponALoop** [comment]
I created a new _fancy_ html template with some filtering using datatables.js. It's available as a [gist here](https://gist.github.com/OnceUponALoop/9e489285eb55a792e3399012d2617854)

![image](https://github.com/anchore/grype/assets/3958336/ec33e318-bb10-48e5-adf2-0fa734b5a74d)

**Comment by spiffcs** [comment]
Wow thanks for the PR @OnceUponALoop! Let me give this a look and we'll include it under our template folder. This is excellent!

**Comment by kston83** [comment]
I made some small changes to @OnceUponALoop's template. I wanted to prioritize the CVE or the GHSA, this version does that and checks for a CVE and CVE severity and displays those, if no CVE data is available it will fall back to the GHSA. The severities get updated accordingly along with links to CVE/GHSA

https://gist.github.com/kston83/e615b7169f7e66156c06d6a6f072f167

**Comment by rydzak115** [comment]
Hi.  I was wondering if anyone has run across this.  I used this template for printing out the locations with an old version of grype.  I am moving to the latest version and it no longer works.  Any help would be greatly appreciated:

"Package","Version Installed","Vulnerability ID","Severity","Vuln. Data Source","Namespace","Fixed State","Fixed Version","RHSA","Locations"
{{- range $match := .Matches}}
"{{$match.Artifact.Name}}","{{$match.Artifact.Version}}","{{$match.Vulnerability.ID}}","{{$match.Vulnerability.Severity}}","{{$match.Vulnerability.DataSource}}","{{$match.Vulnerability.Namespace}}","{{$match.Vulnerability.Fix.State}}","{{$match.Vulnerability.Fix.Versions}}",
{{- range $rhsaid := $match.Vulnerability.Advisories -}} "{{$rhsaid.ID}}"|{{- end}}
{{- range $location := $match.Artifact.Locations -}} "{{$location.RealPath}}"|{{- end}}
{{- end}}

I updated the RHSA printing to the latest version.  I am still not able to print out the Locations, though.

@spiffcs 
final update to this:  What i have learned is that it is imperative to use the syft-json formatted output when creating a csv with the grype template.  Using the cyclone-dx or spdx formats did not work.  but the syft-json format did work with my updated template.

**Comment by samcro1967** [comment]
@rydzak115 It was not working for me either.  Looks like it has not worked in awhile.  I obviously have not been diligent at looking at vulnerabilities in images.  :blush:  I reworked the template and got it working for me again.  Tried to preserve all of the info that was in it previously and get more rows on the screen.  The only thing I wanted, but could not make work, was to allow for advanced filtering so you could filter by Critical, High, and Fixed.  I got it somewhat working, but it looked awful given my limited skills with this stuff.  So, I removed it.  I figured if people really wanted advanced filtering they can export it to Excel and do it there.

**Here is the template file: (remove .txt form the end)**
[html_tmpl.txt](https://github.com/user-attachments/files/21517765/html_tmpl.txt)

**Here is what it looks like:**
<img width="1905" height="858" alt="Image" src="https://github.com/user-attachments/assets/4cb2d91a-aa63-4ca4-ac6a-c02fd1f894df" />

**Comment by samcro1967** [comment]
I should also note that I failed to get the severity and state column values colored to match the header total colors.  I tried all of the following:

- Custom CSS
- columnDefs.render in JavaScript
- Custom HTML in Template (RowTemplate)

DataTables was just fighting me the whole way and not allowing me to override its defaults in the table.  If anyone knows a way to do that, let me know and I will take another look at it.  Would be a nice visual to have I think.

**Comment by OnceUponALoop** [comment]
@samcro1967 i used pills to get the coloring to match, I recently submitted a PR (#2635) for the html template to add a bunch of new functionality, have a look

**Comment by samcro1967** [comment]
@OnceUponALoop Thank you very much.  I thought I pulled the most recent html.tmpl and got CORS errors on some of the link and script imports.  Not sure how I missed yours.  I was very close to getting the pills working, but was able to get it across the finish line with your help.  Love the app icons and expandable sections also.  I incorporated them into my template and cited creds for you in the doc string.  Allowed me to get some more screen space to display more data and still easily keep 10 rows on a laptop screen.  Appreciate the assist.

[samcro1967.tmpl.txt](https://github.com/user-attachments/files/21537733/samcro1967.tmpl.txt)

<img width="1461" height="751" alt="Image" src="https://github.com/user-attachments/assets/79e42329-57da-4292-9cc6-508aaec5b187" />
